kylin-spark-project/kylin-spark-common/src/main/spark31/org/apache/spark/sql/execution/KylinJoinSelection.scala

```diff
@@ -0,0 +1,283 @@
+ * side with the broadcast hint. If neither side has a hint, we broadcast the side with
+ * the smaller estimated physical size.
+ */
+case class KylinJoinSelection(session: SparkSession) extends Strategy with PredicateHelper with Logging {
+
+  val conf: SQLConf = session.sessionState.conf
+
+  /**
+   * Matches a plan whose output should be small enough to be used in broadcast join.
+   */
+  private def canBroadcast(plan: LogicalPlan): Boolean = {
+    val sizeInBytes = plan.stats.sizeInBytes
+    sizeInBytes >= 0 && sizeInBytes <= conf.autoBroadcastJoinThreshold && JoinMemoryManager.acquireMemory(sizeInBytes.toLong)
+  }
+
+  /**
+   * Matches a plan whose single partition should be small enough to build a hash table.
+   *
+   * Note: this assume that the number of partition is fixed, requires additional work if it's
+   * dynamic.
+   */
+  private def canBuildLocalHashMap(plan: LogicalPlan): Boolean = {
+    plan.stats.sizeInBytes < conf.autoBroadcastJoinThreshold * conf.numShufflePartitions
+  }
+
+  /**
+   * Returns whether plan a is much smaller (3X) than plan b.
+   *
+   * The cost to build hash map is higher than sorting, we should only build hash map on a table
+   * that is much smaller than other one. Since we does not have the statistic for number of rows,
+   * use the size of bytes here as estimation.
+   */
+  private def muchSmaller(a: LogicalPlan, b: LogicalPlan): Boolean = {
+    a.stats.sizeInBytes * 3 <= b.stats.sizeInBytes
+  }
+
+  private def canBuildRight(joinType: JoinType): Boolean = joinType match {
+    case _: InnerLike | LeftOuter | LeftSemi | LeftAnti | _: ExistenceJoin => true
+    case _ => false
+  }
+
+  private def canBuildLeft(joinType: JoinType): Boolean = joinType match {
+    case _: InnerLike | RightOuter => true
+    case _ => false
+  }
+
+  private def broadcastSide(
+                             canBuildLeft: Boolean,
+                             canBuildRight: Boolean,
+                             left: LogicalPlan,
+                             right: LogicalPlan): BuildSide = {
+
+    def smallerSide =
+      if (right.stats.sizeInBytes <= left.stats.sizeInBytes) BuildRight else BuildLeft
+
+    if (canBuildRight && canBuildLeft) {
+      // Broadcast smaller side base on its estimated physical size
+      // if both sides have broadcast hint
+      smallerSide
+    } else if (canBuildRight) {
+      BuildRight
+    } else if (canBuildLeft) {
+      BuildLeft
+    } else {
+      // for the last default broadcast nested loop join
+      smallerSide
+    }
+  }
+
+  private def canBroadcastByHints(joinType: JoinType, left: LogicalPlan, right: LogicalPlan, hint: JoinHint)
+  : Boolean = {
+//    val buildLeft = canBuildLeft(joinType) && logical.BROADCAST.equals(hint.leftHint.get.strategy.get)
+//    val buildRight = canBuildRight(joinType) && logical.BROADCAST.equals(hint.rightHint.get.strategy.get)
+//    buildLeft || buildRight
+    false
```

Comment:
Why comments these lines and return false directly ?

Reply:
@zzcclp fixed